### DISCUSSION TITLE
Define a standard filesystem library

### DISCUSSION THREAD

**Discussion by jonasraoni** [pkp/pkp-lib#8776] [Accepted Proposals]
## Problem

At this moment the system makes use of at least 3 types of filesystem calls:
- PHP filesystem
- FileManager class
- [Flysystem](https://flysystem.thephpleague.com/)

Also, we've got some generic settings, such as the `[files].umask` that might be forgotten.

## Ideas/Suggestions
- The standard PHP functions should be avoided, checking the success result is error-prone, and when forgotten, can fail silently. Also, they generate log pollution.
- `Flysystem` is probably a good contender, as it emits `Exceptions` on failures and we're already using it partially, it does support setting a default file mask.
- Some functions from the `FileManager` might be interesting to keep around
- It sounds interesting to have a `Flysystem` instance for specific places (files, public, application), probably kept at the container, for security purposes (it has protection against navigations `../../..`).

**Comment by asmecher** [comment]
@jonasraoni, this is a known work in progress -- we have decided to adopt Flysystem for filesystem interaction, as per the submission file tracking stuff, but won't be doing the migration big-bang style. As components of the system relating to the filesystem are rewritten, they should adopt Flysystem. (I don't mind if this is never 100% completed, as long as we're working over time to remove the more homegrown aspects of our codebase.) Is there anything still needing to be figured out?

**Reply by jonasraoni** [reply]
If it's already decided, then it's just needed to close this discussion and watch out the PRs :D
